src/ResultSetBuilder.php

```diff
@@ -35,34 +28,34 @@ private function buildResults(Response $response): array
 
         foreach ($data['hits']['hits'] as $hit) {
             $result = new Result($hit);
-            $result->setModel($this->buildModel($result));
+            $result->setModel($this->buildModelResult($result));
             $results[] = $result;
         }
 
-        return $results;
+        return new ResultSet($response, $query, $results);
     }
 
-    private function buildModel(Result $result)
+    public function buildModelFromIndexAndData(string $indexName, $data)
     {
-        $source = $result->getSource();
+        $class = $this->client->getClassFromIndexName($this->client->getPureIndexName($indexName));
 
-        if (empty($source)) {
-            return null;
-        }
+        $context = $this->client->getSerializerContext($class);
 
-        return $this->buildModelFromIndexAndData($result->getIndex(), $source);
+        return $this->client->getSerializer()->denormalize($data, $class, null, $context);
     }
 
-    public function buildModelFromIndexAndData($indexName, $data)
+    private function buildModelResult(Result $result)
     {
-        $pureIndexName = $this->client->getPureIndexName($indexName);
-        $indexToClass = $this->client->getConfig(Client::CONFIG_INDEX_CLASS_MAPPING);
-        if (!isset($indexToClass[$pureIndexName])) {
-            throw new RuntimeException(sprintf('Unknown class for index "%s", did you forgot to configure "%s"?', $pureIndexName, Client::CONFIG_INDEX_CLASS_MAPPING));
+        $source = $result->getSource();
+        if (!$source) {
+            return null;
         }
 
-        $context = $this->client->getSerializerContext($indexToClass[$pureIndexName]);
+        $class = $this->client->getClassFromIndexName($this->client->getPureIndexName($result->getIndex()));
+
+        $context = $this->client->getSerializerContext($class);
+        $context['result'] = $result;
```

Comment:
Maybe add a prefix or something? `elastically_result`? To avoid any side effect on existing context.

Reply:
& add a const in the class to define the context key (like the Serializer does for example) if possible

Reply:
thanks, fixed